src/lib/transition_handler/validator.ml

```diff
@@ -0,0 +1,82 @@
+open Core_kernel
+
+module Make (Inputs : Inputs_intf.S) = struct
+  open Inputs
+  open Consensus_mechanism
+
+  let validate_transition frontier t =
+    let log_assert condition error_msg =
+      let log () = Logger.info logger "transition rejected: %s" error_msg; false in
+      condition || log ()
+    in
+    let consensus_state = Fn.compose Protocol_state.consensus_state External_transition.protocol_state in
+    let root = Transition_frontier.Root.transition (Transition_frontier.root frontier) in
+    log_assert
+      (Consensus_mechanism.is_valid (consensus_state t))
+      "failed consensus validation"
+    && log_assert
+      (Consensus_mechanism.select (consensus_state root) (consensus_state t) = `Take)
+      "was not better than transition frontier root";
+    && log_assert
+      (Proof.check (External_transition.proof t))
+      "proof was invalid"
+
+  let run ~transition_reader ~valid_transition_writer frontier =
+    don't_wait_for (Linear_pipe.Reader.iter ~f:(fun transition ->
+        if validate_transition frontier transition then
+          Linear_pipe.write valid_transition_writer transition
```

Comment:
We said this should be drop-old, not blocking:
https://github.com/CodaProtocol/coda/blob/master/docs/res/transition_frontier_controller.png

I still think it should be drop old (basically this would happen when validation takes longer than the rate at which messages were coming in)